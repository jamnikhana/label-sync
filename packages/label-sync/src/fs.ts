import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

import { Maybe } from './utils'

const fsReaddir = promisify(fs.readdir)

/**
 * Looks up for a folder starting with the current directory and
 * going up until it reaches the root folder or finds the folder.
 *
 * @param dir
 * @param name
 */
export async function findFolderUp(
  dir: string,
  name: string,
): Promise<Maybe<string>> {
  return findUp(dir, p => path.dirname(p) === name)
}

/**
 * Traverses file system up to find a specific directory that contains
 * a matching pattern.
 *
 * @param dir
 * @param pattern
 */
export async function findUp(
  dir: string,
  pattern: (path: string) => boolean,
): Promise<Maybe<string>> {
  switch (path.normalize(dir)) {
    /* End case: we reached the root. */
    case '/': {
      return null
    }
    /* Recursive case. */
    default: {
      const elements = await fsReaddir(dir)
      const includes = elements.some(pattern)

      if (includes) return dir
      else return findUp(path.join(dir, '../'), pattern)
    }
  }
}
