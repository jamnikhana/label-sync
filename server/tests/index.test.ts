import fs from 'fs'
import nock from 'nock'
import path from 'path'
import { Probot } from 'probot'

import ls from '../src'

/* Fixtures */
const configFixture = fs.readFileSync(
  path.resolve(__dirname, './__fixtures__/labelsync.yml'),
  { encoding: 'utf-8' },
)
import labelCreatedPayload from './__fixtures__/label.created'
import pushPayload from './__fixtures__/push'

describe('bot:', () => {
  beforeAll(() => {
    nock.disableNetConnect()
  })

  afterAll(() => {
    nock.enableNetConnect()
  })

  let probot: Probot

  beforeEach(() => {
    probot = new Probot({})
    const app = probot.load(ls)

    app.app = {
      getSignedJsonWebToken: () => 'jwt',
      getInstallationAccessToken: () => Promise.resolve('token'),
    }
  })

  test(
    'push event',
    async () => {
      expect.assertions(2)

      const configEndpoint = jest.fn().mockReturnValue({
        content: Buffer.from(configFixture).toString('base64'),
      })

      const installationsEndpoint = jest.fn().mockReturnValue({
        repositories: [
          { name: 'graphql-shield' },
          { name: 'emma-cli' },
          { name: 'nookies' },
        ],
      })

      const labelsEndpoint = jest.fn().mockImplementation(uri => {
        switch (uri) {
          case '/repos/maticzav/graphql-shield/labels': {
            return [
              {
                name: 'bug/0-needs-production',
                color: 'ff0022',
              },
              {
                name: 'bug/1-has-reproduction',
                color: 'ff0022',
                description: 'Indicates that an issue has reproduction',
              },
              {
                name: 'bug/2-bug-confirmed',
                color: 'blue',
              },
            ]
          }
          case '/repos/maticzav/emma-cli/labels': {
            return [
              {
                name: 'bug/0-needs-production',
                color: 'ff0022',
              },
              {
                name: 'bug/1-has-reproduction',
                color: 'ff0022',
                description: 'Indicates that an issue has reproduction',
              },
              {
                name: 'bug/2-bug-confirmed',
                color: 'red',
              },
              {
                name: 'bug/3-fixing',
                color: '00ff22',
                description:
                  'Indicates that we are working on fixing the issue.',
              },
            ]
          }
          default: {
            throw new Error(`Unknown uri: ${uri}`)
          }
        }
      })

      /* Mocks */

      nock('https://api.github.com')
        .post('/app/installations/13055/access_tokens')
        .reply(200, { token: 'test' })

      nock('https://api.github.com')
        .get(
          '/repos/maticzav/maticzav-labelsync/contents/labelsync.yml?ref=refs%2Fheads%2Fmaster',
        )
        .reply(200, configEndpoint)

      nock('https://api.github.com')
        .get('/installation/repositories?per_page=100')
        .reply(200, installationsEndpoint)

      nock('https://api.github.com')
        .get(/\/repos\/maticzav\/.*\/labels/)
        .reply(200, labelsEndpoint)

      nock('https://api.github.com')
        .post(/\/repos\/maticzav\/.*\/labels/)
        .reply(200, (uri, body) => {
          switch (uri) {
            case '/repos/maticzav/graphql-shield/labels': {
            }
            case '/repos/maticzav/emma-cli/labels': {
            }
          }
        })

      nock('https://api.github.com')
        .delete(/\/repos\/maticzav\/.*\/labels/)
        .reply(200, (uri, body) => {
          switch (uri) {
            case '/repos/maticzav/graphql-shield/labels': {
            }
            case '/repos/maticzav/emma-cli/labels': {
            }
          }
        })

      await probot.receive({ id: 'push', name: 'push', payload: pushPayload })

      /* Tests */

      expect(configEndpoint).toBeCalledTimes(1)
      expect(installationsEndpoint).toBeCalledTimes(1)
      expect(labelsEndpoint).toBeCalledTimes(2)
    },
    5 * 60 * 1000,
  )

  test(
    'push event with insufficient premissions',
    async () => {
      expect.assertions(3)

      const configEndpoint = jest.fn().mockReturnValue({
        content: Buffer.from(configFixture).toString('base64'),
      })

      const installationsEndpoint = jest.fn().mockReturnValue({
        repositories: [{ name: 'graphql-shield' }],
      })

      /* Mocks */

      nock('https://api.github.com')
        .post('/app/installations/13055/access_tokens')
        .reply(200, { token: 'test' })

      nock('https://api.github.com')
        .get(
          '/repos/maticzav/maticzav-labelsync/contents/labelsync.yml?ref=refs%2Fheads%2Fmaster',
        )
        .reply(200, configEndpoint)

      nock('https://api.github.com')
        .get('/installation/repositories?per_page=100')
        .reply(200, installationsEndpoint)

      nock('https://api.github.com')
        .post('/repos/maticzav/maticzav-labelsync/issues')
        .reply(200, body => {
          expect(body).toMatchSnapshot()
        })

      await probot.receive({
        id: 'pushinsufficient',
        name: 'push',
        payload: pushPayload,
      })

      /* Tests */

      expect(configEndpoint).toBeCalledTimes(1)
      expect(installationsEndpoint).toBeCalledTimes(1)
    },
    5 * 60 * 1000,
  )

  test.skip('label.created event removes unsupported labels', async () => {
    expect.assertions(3)

    const configEndpoint = jest.fn().mockReturnValue({
      content: Buffer.from(configFixture).toString('base64'),
    })

    const installationsEndpoint = jest.fn().mockReturnValue({
      repositories: [{ name: 'graphql-shield' }],
    })

    /* Mocks */

    nock('https://api.github.com')
      .post('/app/installations/13055/access_tokens')
      .reply(200, { token: 'test' })

    nock('https://api.github.com')
      .get(
        '/repos/maticzav/maticzav-labelsync/contents/labelsync.yml?ref=refs%2Fheads%2Fmaster',
      )
      .reply(200, configEndpoint)

    nock('https://api.github.com')
      .delete('/repos/maticzav/maticzav-labelsync/issues')
      .reply(200, (uri, body) => {
        console.log(uri, body)
      })

    await probot.receive({
      id: 'labelcreated',
      name: 'label.created',
      payload: labelCreatedPayload,
    })

    /* Tests */

    expect(configEndpoint).toBeCalledTimes(1)
    expect(installationsEndpoint).toBeCalledTimes(1)
  })
})
