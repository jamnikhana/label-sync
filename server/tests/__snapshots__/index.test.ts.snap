// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bot: pull_request 1`] = `
Object {
  "body": "## PullRequest Overview 
> NOTE: This is only a preview of what will happen. Nothing has been changed yet.

#### \\\\\`graphql-shield\\\\\`

__New labels:__
 * bug/3-fixing

__Changed labels:__
 * bug/2-bug-confirmed

__Removed labels:__
 * bug/4-fixed
#### \\\\\`emma-cli\\\\\`

__New labels:__
No new labels created.

__Changed labels:__
No changed labels.

__Removed labels:__
You haven't removed any label.",
}
`;

exports[`bot: pull_request with insufficient permissions 1`] = `
Object {
  "body": "It seems like this configuration stretches beyond
repositories we can access. Please update it so we can help
you as best as we can.

_Missing repositories:_
 * emma-cli",
}
`;

exports[`bot: push event with insufficient premissions 1`] = `"/repos/maticzav/maticzav-labelsync/issues"`;
